### PR TITLE
#11036: CSV: enable (command-)button only after client side validation passes (ui-state-error + ui-state-valid)

### PR DESCRIPTION
**Pull request by christophs78**
Just started working on this. Will probably take at least weeks maybe even a month or tow.

### CODE REVIEW

**Comment by tandraschko** [comment]
can you explain a bit the mechanism?

**Comment by christophs78** [comment]
> can you explain a bit the mechanism?

Basic idea is to plug into following events:

- full page load (not yet implemented)
- ajax-update (not yet implemented)
- `p:clientValidator` - event

Whenever one of this events is triggered we look for instances of `p:commandButton` marked with `enabledByValidateClient="true"`. Each `p:commandButton` marked such way has following data-attributes: `data-pf-validateclient-ajax="true" data-pf-validateclient-process="j_idt286:selectOneMenu" data-pf-validateclient-update="j_idt286:grid"`

We use this information / metadata to run CSV and decide afterwards whether to enable or disable the button.

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/csv/Csv002Test.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        // Assert
+        Assertions.assertEquals("", page.msgSelectBooleanCheckbox.getText());
+
+        // Act
+        page.selectBooleanCheckbox.check();
+        page.selectBooleanCheckbox.uncheck();;
+
+        // Assert
+        Assertions.assertEquals("yes/no: Validation Error: Value is required.",
```
Comment (christophs78) [inline primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/csv/Csv002Test.java]: Any idea what CSV should do with `p:selectBooleanCheckbox` and `p:triStateCheckbox` set to required?
image
IMO it does currently nothing.

(I´m adding CSV-IT´s as part of this PR too. This IT fails currently.)

**Comment by christophs78** [comment]
As usual there are more things we may do.
But IMO we are at a good point to review this PR. Additional i provided our first IT´s for CSV as part of this PR. (Two of them disabled for now because required="true" does not seem to work (make sense?) for some components like `SelectBooleanCheckbox` and `TriStateCheckbox`.)

**Comment by melloware** [comment]
I know required doesn't make sense for Select Boolean Checkbox but Tristate one might make sense?

**Comment by christophs78** [comment]
Yep required=true makes IMO sense for TriStateCheckbox. Even for SelectBooleanCheckbox there may be some usecases where it makes sense. e.g. you need to check this checkbox to continue during some registration form. But there may be usecases where it does not make sense. (No one forces you to use required=true, so there may be room to implement required=true even for SelectBooleanCheckbox )
IMO we should open separate issues for this once we figured out what´s "common sense".

**Comment by melloware** [comment]
Agreed

**Comment by tandraschko** [comment]
do we really need additional css classes? styles should be the same, indepdent of CSV and SSV. If you need to know if it was triggered by CSV, maybe add a data attr?

**Comment by christophs78** [comment]
> do we really need additional css classes? styles should be the same, indepdent of CSV and SSV. If you need to know if it was triggered by CSV, maybe add a data attr?

In an earlier iteration i had data-attributes. Then i refactored it to styles classes.
Our current style-classes only cover the invalid-state (`ui-state-error`) but IMO we want/need also style-classes for the valid state. So i took some inspiration from https://angular.io/guide/form-validation#control-status-css-classes.

My target was to allow something like following with JSF/PrimeFaces
image

An option would be to stick with `ui-state-error` and add something new like `ui-state-valid`. Implicit knowing this is only set (at least for now) via CSV.

**Comment by tandraschko** [comment]
+1 for ui-state-valid for now
if you need to know that it failed/succeeded in CSV, we can still later check the best solution

**Comment by christophs78** [comment]
Behaviour before i start reworking `ui-state-xxx` css-classes.
https://github.com/primefaces/primefaces/assets/10461942/60e4e9c2-311a-4620-853d-4c062d760172

**Comment by christophs78** [comment]
After CSS-class rework (reuse `ui-state-error` + introduce `ui-state-valid` instead of `ui-state-csv-valid` + `ui-state-csv-invalid`)
I personally dislike the red borders we get this way at inital loading.
https://github.com/primefaces/primefaces/assets/10461942/d51121b5-8620-415d-a8d6-0c51b4b4677c

**Comment by christophs78** [comment]
> @christophs78 but if its `ui-state-error` doesn't it have to have those borders?

Yep. From a UX-point-of-view i´d like to rollback the last commit. (back to `ui-state-csv-valid` + `ui-state-csv-invalid` and have more flexibility how to style this)
I hope it´s now transparent what our trade-off´s are. A bit of UX vs new/additional CSS-classes which may cause a bit of confusion elsewhere.)

File: primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandbutton.js [hunk 2]
```diff
@@ -23,6 +25,18 @@ PrimeFaces.widget.CommandButton = PrimeFaces.widget.BaseWidget.extend({
         PrimeFaces.skinButton(this.jq);
 
         this.bindTriggers();
+
+        if (cfg.validateClientDynamic) {
+            let that = this;
+
+            // init enabled/disabled-state after initial page-load
+            setTimeout( function() {
```
Comment (tandraschko) [inline primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandbutton.js]: should we use the new task api @melloware introduced?

**Comment by tandraschko** [comment]
so you would like to introduce a new style for csv-invalid and csv-valid? this means we need to change the themes and themecreator propably too

does PrimeVue has such a feature? can you compare style classes?

File: primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandbutton.js (hunk 2 above)
Comment (melloware) [reply primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandbutton.js]: Yes please use `PrimeFaces.queueTask`

Comment (christophs78) [reply primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandbutton.js]: Thanks for pointing me on this. Reminded there was some replacement for window.setTimeout but could not figure out myself what anymore.

**Comment by christophs78** [comment]
> so you would like to introduce a new style for csv-invalid and csv-valid? this means we need to change the themes and themecreator propably too
> 
> does PrimeVue has such a feature? can you compare style classes?

The new css class(es) - ui-state-valid respective ui-state-csv-valid + ui-state-csv-invalid are IMO a additional offering which allow you to do some project-specific styling. Does not necessary mean we need to add something to themes and themecreator.

I´ll try to look what the other Prime... - libraries do.

**Comment by christophs78** [comment]
What do other Prime-libraries?

- https://primeng.org/inputtext#invalid
- https://primevue.org/inputtext/#invalid

File: primefaces/src/main/resources/META-INF/resources/primefaces/validation/validation.common.js [hunk 3]
```diff
@@ -511,6 +550,36 @@ if (window.PrimeFaces) {
...
+         * Bind to Ajax-Complete-events to update CSV-state after an Ajax-call may have changed state.
+         * @internal
+         */
+        bindAjaxComplete: function() {
+            if (this.ajaxCompleteBound) return;
+
+            var doc = $(document),
+                $this = this;
+
+            doc.on('pfAjaxComplete', function(e, xhr, settings, args) {
```
Comment (melloware) [inline primefaces/src/main/resources/META-INF/resources/primefaces/validation/validation.common.js]:
```suggestion
            $(document).on('pfAjaxComplete', function(e, xhr, settings, args) {
```

No need to store `doc` variable as its not used anywhere else?

**Review by melloware — CHANGES_REQUESTED** [review]

File: primefaces/src/main/resources/META-INF/resources/primefaces/validation/validation.common.js (hunk 3 above)
Comment (christophs78) [reply primefaces/src/main/resources/META-INF/resources/primefaces/validation/validation.common.js]: good catch - thanks!

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/csv/Csv002Test.java (hunk 1 above)
Comment (christophs78) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/csv/Csv002Test.java]: `@AssertTrue` helps (partial) for SelectBooleanCheckbox.

**Comment by christophs78** [comment]
Do we miss something on this PR?
When it helps i may provide another PR for the other CSS-variante. (Should be easy to provide this mutation.)
IMO both variants are backward-compatible. There are simple some new additional css classes which you may use in your project for custom styling when you want.

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
If @tandraschko is ok with it we can get it in for the 14 RC1 release.

**Comment by christophs78** [comment]
State after loading:
(ui-state-error kicks in immediately)
image

State after some inputs (ui-state-valid is added after successful CSV)
image

**Pull request closed without merging**